### ISSUE TITLE
API: Change normalized=False to normalize=True in Rotation

### ISSUE THREAD

**Issue by nmayorov** [scipy/scipy#10968]
In `Rotation` we used keyword `normalized` to indicate that normalization of vectors/quaternions was done:

`def __init__(self, quat, normalized=False, copy=True)` So if `normalized=False`, then normalization is required.

I think it was an unfortunate choice and passing imperatives as `copy`, `normalize`, etc. is much more common, natural and easier to reason about. 

There are two public methods with `normalized`: `from_quat` and `match_vectors`. The `__init__` might be considered as private.

I think ideally we change that, but I'm not sure how to do it in a straightforward and safe way through a deprecation cycle.

Is it worth it? If yes, how can we do it?

@rgommers @matthew-brett @pmla

**Comment by pmla** [comment]
Sorry for a slow reply - I overlooked this despite being pinged.

I strongly agree with the proposal -  `normalized` is always requires me to think twice about what I'm doing. I propose that we remove the `normalized` keyword from the public methods entirely.

For `from_quat` we can create a private `_from_quat` method with a `normalize` parameter. Let the public `from_quat` method call this function with `normalize=True`. The overhead of normalization is quite small anyway.

For `match_vectors`, we could just let the user normalize the inputs if needed. Is this even a common use case? (A genuine question - I don't know)

In both public methods the `normalized` parameter is the last parameter, so it should be easy to deprecate using `*args` and `**kwargs`.

**Comment by nmayorov** [comment]
@pmla 

Thanks for the feedback! I have an idea how to gradually move from `normalized` to `normalize`:
```Python
@classmethod
def from_quat(cls, quat, normalized=None, normalize=True):
    if normalized is not None:
        # issue deprecation, use normalized
    else:
       # use normalize       
```
It's a bit convoluted, but will do the job. 

OK, I'll prepare these changes, maybe someone else will look at them. It's better to do in 1.4.0.

**Comment by pmla** [comment]
Nice. This is simpler and better than my proposal.

**Comment by nmayorov** [comment]
Now looking at it, I think it might be better to simply deprecate `normalized` (and then remove) and stick to some fixed strategy --- never normalize (expect normalized input) or always normalize.

What do you think about this approach? Which "fixed strategy" is more logical if we go this way?

**Comment by pmla** [comment]
In that case I would go with:
* Always normalize input to `from_quat`.
* Never normalize input to `match_vectors`.

We can keep the `normalize` parameter for use within the class though.

**Comment by nmayorov** [comment]
> Never normalize input to match_vectors.

Well, I think the computations are cheap and the problem is always formulated for unit vectors. I think it's not bad to enforce the unit norms. What do you think?

**Comment by pmla** [comment]
> the problem is always formulated for unit vectors

Oh - I hadn't noticed that the method is defined for unit vectors only. This is an unnecessary restriction. The rotational alignment problem is very commonly used in e.g. structural biology to align two molecules. It is quite easy to implement the general case.

A classic paper on the problem:
Closed-form solution of absolute orientation using unit quaternions
Berthold K. P. Horn

I think we should make this method handle the general case, after which it makes sense to never normalize the input vectors.

**Comment by nmayorov** [comment]
> Oh - I hadn't noticed that the method is defined for unit vectors only. This is an unnecessary restriction. The rotational alignment problem is very commonly used in e.g. structural biology to align two molecules. It is quite easy to implement the general case.

Yes, I was thinking about that as well. But in your example --- do we need to have at least equal length vectors? Does it make sense to match arbitrary length vectors? In applications I know vectors are normalized to unit norm to remove this "useless" degree of freedom.

**Comment by pmla** [comment]
> Does it make sense to match arbitrary length vectors?

Yes - it will affect the calculated rotation. Consider two sets of unit vectors, A and B. Calculate a optimal rotation. Now take of the vectors in A and make it a million times longer. Now it carries more weight in determining the optimal rotation.

**Comment by nmayorov** [comment]
And if corresponding vectors don't have equal lengths? Should it be considered as bad problem formulation?

**Comment by pmla** [comment]
No, this is a normal formulation. It just means that the alignment has non-zero root-mean-square deviation (RMSD).

**Comment by nmayorov** [comment]
It sounds convincing, I'll remove the normalization then with some additional explanation.


---- On Tue, 12 Nov 2019 18:18:39 +0500 notifications@github.com wrote ----



No, this is a normal formulation. It just means that the alignment has non-zero root-mean-square deviation (RMSD).

**Comment by pmla** [comment]
I will have to read Markley's paper ([2]) to understand the sensitivity matrix. I'm not sure if it will be affected by non-unit vectors. I also think we should replace ref. 1 with a more general description, perhaps just with a link to the Wikipedia article on the Kabsch algorithm: https://en.wikipedia.org/wiki/Kabsch_algorithm
The Kabsch algorithm is more both general and predates Markley's work by a decade or so.

**Comment by nmayorov** [comment]
> I will have to read Markley's paper ([2]) to understand the sensitivity matrix. I'm not sure if it will be affected by non-unit vectors.

I empirically checked the properties of the returned sensitivity. 

When I increased the length of all vectors by `10`, the returned sensitivity is decreased by `100`. It means that we should multiply it by linear perturbation variance (not angular). 

When I increased the length of half of the vectors by `10`, the return sensitivity was about 2x higher than in the previous case, because we use half as less "strong" observations. All in all it looks reasonable.

As for vectors of strongly unmatched lengths --- I can guess that the contribution of such pairs is as if they have the same length of `sqrt(l1 * l2)`.

I'll try to updated documentation.
